src/util/LazyJsonParser.cpp

```diff
@@ -0,0 +1,159 @@
+#include "LazyJsonParser.h"
+
+#include <absl/strings/ascii.h>
+#include <absl/strings/str_cat.h>
+#include <absl/strings/str_join.h>
+
+namespace ad_utility {
+
+LazyJsonParser::LazyJsonParser(std::vector<std::string> arrayPath)
+    : arrayPath_(arrayPath),
+      prefixInArray_(absl::StrCat(
+          absl::StrJoin(arrayPath.begin(), arrayPath.end(), "",
+                        [](std::string* out, const std::string& s) {
+                          absl::StrAppend(out, "{\"", s, "\": ");
+                        }),
+          "[")),
+      suffixInArray_(absl::StrCat("]", std::string(arrayPath.size(), '}'))) {}
+
+std::string LazyJsonParser::parseChunk(const std::string& inStr) {
+  size_t idx = input_.size();
+  absl::StrAppend(&input_, inStr);
+  size_t materializeEnd = 0;
+
+  auto tryAddKeyToPath = [this]() {
+    if (strEnd_ != 0) {
+      curPath_.emplace_back(input_.substr(strStart_, strEnd_ - strStart_));
+    }
+  };
+
+  if (inString_) {
+    parseString(idx);
+    ++idx;
```

Comment:
As these functions are called both from out- and inside the index-incrementing for-loop this is required. Otherwise the for-loop would skip the char after the consumed part after function-calls from within the loop.

However i see that this at least has to be explained in a comment.